### DISCUSSION TITLE
ObservableProperty code not being generated

### DISCUSSION THREAD

**Discussion by CADTraveller** [CommunityToolkit/dotnet#308] [Q&A] [answered]
I'm trying to utilize this toolkit in a WPF 4.7.2 application.  I have brought in CommunityToolkit.Mvvm V8.0.0 (preview 4).  The I had my main VM inherit from ObservableObject:

```
using CommunityToolkit.Mvvm.ComponentModel;

namespace hsbWPF.Revit.UI.ViewModels
{
    public partial class ParametersMainViewModel : ObservableObject
    { ...
```

However, the 'partial' declaration is dim since there is no other partial class.  When I mark a field with the [ObservableProperty] attribute there is no matching Public property generated. Nothing shows up in the Analyzers section of References:

![image](https://user-images.githubusercontent.com/5366736/173119504-ea4b0ce2-1fe8-4365-8902-1eac09a67018.png)

Is there a limitation around C# or Framework versions?  

What else can I check?

**Comment by Sergio0694** [comment]
Just to double check, is the code just not showing up there or is it just not being generated at all?
As in, if you try to use the generated property and build, does it fail to build?
Do you see any errors in the output window?

**Reply by CADTraveller** [reply]
Thanks for the follow up Sergio.

It appears that nothing is being generated at all.  If I mark a field in my ObservableObject class with the ObservablePropety attribute:

```
[ObservableProperty]
        private string statusMessage;
```

The Property StatusMessage is not available and cannot be used; it has not been generated.

If I declare StatusMessage as a standard Property then the Project builds with no errors.  I can use the SetProperty(ref statusMessage, value) method.   

I see only typical Warnings in Output from my WIP code, no errors, and nothing specifcally from the Toolkit.

**Reply by Sergio0694** [reply]
I wonder if this is the same WPF-related issue that has come up a while back. Can you try moving the code to a separate .NET Standard (or .NET library, whatever you're using in your app) project, and reference that from your WPF app? The generator should work there.

**Reply by CADTraveller** [reply]
A bit of a messy refactoring to move my ViewModels to .NET Standard, but of course it offers some advantages outside of troubleshooting this issue.

Working on it now, will report back the results.

**Reply by CADTraveller** [reply]
I haven't gotten to the point where I can actually run this test yet; my Project had extensive use of ICollectionView which will have to be refactored to some other type to become .NET Standard.

However, even at this stage I seem to be in better shape. Code generation is creating classes, which it wasn't doing in the WPF 4.7.2 Project:

![image](https://user-images.githubusercontent.com/5366736/173585244-8e9e50f3-d2c5-4e50-b3af-aa738571b43d.png)

It seems you were correct @Sergio0694

**Reply by CADTraveller** [reply]
Ah, no. I think I found the heart of the issue. My projects are C# 7.3.

_Error	MVVMTK0008	The source generator features from the MVVM Toolkit require consuming projects to set the C# language version to at least C# 8.0	Revit.PropertyData	X:\HsbDefault\BeamApp\Hsb_NetApi\hsbWPF.Properties.UI\Revit.PropertyData\CSC	1_	

To work with the rest of our internal code I'm limited to .NET Standard 2.0.  So, the Toolkit is not applicable for me currently.

Thanks for your assistance @Sergio0694

**Reply by Sergio0694** [reply]
I'm confused, weren't you getting that diagnostic error before? It should've showed up, no?
Also, FYI you can manually set the language version to C# 8.0 if you want: add `<LangVersion>8.0</LangVersion>` to your .csproj.

**Reply by CADTraveller** [reply]
I had not seen that error message in my original WPF project. 

I know I can manually set the C# version, but .NET Standard 2.0 doesn't support C# 8.0 does it?

**Reply by Sergio0694** [reply]
> "I had not seen that error message in my original WPF project."

Mmh that's not good, we should still be able to show that diagnostic. I wonder if that's just a tooling issue 🤔
I'll try to see if I can repro. Just to confirm: is your WPF project using an SDK-style .csproj, or old-style? What TFM are you using?

>"but .NET Standard 2.0 doesn't support C# 8.0 does it?"

Eh, good question. Officially, no. But, you can bump the version manually and most things will still work. Some will require some additional work. Some will just not work, but you can just not use those. I've written an overview [here](https://sergiopedri.medium.com/enabling-and-using-c-9-features-on-older-and-unsupported-runtimes-ce384d8debb), if it helps 🙂

**Reply by CADTraveller** [reply]
My WPF project is Framework 4.7.2 with the old style csproj file.  Working with a bunch of legacy code and doing libraries for AutoCAD and Revit, we're kind of constrained...

Nice article @Sergio0694 , I'd found a similar one from James M. but yours is a bit newer. And indeed, adding the <LangVersion> element worked!  Code is generated as expected.  I still need to do some integration testing but this issue is closed.

Thanks again for your prompt response and useful feedback ;-]

**Comment by CADTraveller** [comment, marked as answer]
Moving the ViewModels to a .NET Standard library that uses C# 8.0 enabled the Toolkit Code Generation.
